File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        //optimal approach: 
        int mx_profit = 0;
        int min_price = INT_MAX;
        
        for(int i = 0;i<prices.size();i++){
            min_price = min(min_price,prices[i]);
            mx_profit = max(mx_profit,prices[i]-min_price);
        }
        return mx_profit;
    }
};

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // Approach_02:better approch:time:O(n) and space:O(1)
        int maxProfit = Integer.MIN_VALUE;
        int minPrice = Integer.MAX_VALUE;
        
        for(int i = 0;i<prices.length;i++)
        {
            minPrice = Math.min(minPrice, prices[i]);
            maxProfit = Math.max(maxProfit, prices[i]-minPrice);
        }
        return maxProfit;
    }
}